pkg/providers/instanceprofile/types.go

```diff
@@ -0,0 +1,73 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package instanceprofile
+
+import (
+	"strings"
+
+	"github.com/aws/smithy-go"
+	cache "github.com/patrickmn/go-cache"
+	"github.com/samber/lo"
+)
+
+type RoleNotFoundError struct {
+	error
+}
+
+func IsRoleNotFoundError(err error) bool {
+	_, ok := lo.ErrorsAs[RoleNotFoundError](err)
+	return ok
+}
+
+// ToRoleNotFoundError converts a smithy.APIError returned by AddRoleToInstanceProfile to a RoleNotFoundError.
+func ToRoleNotFoundError(err error) (error, bool) {
```

Comment:
Why do we need a custom error type here vs. just using the error as-is if we are only storing role not found errors in this cache anyways?

Reply:
Good point. I had previously made this a custom error because I was going to check for it in the instance profile reconciler and set a longer `RequeueAfter` period, but that would have meant I couldn't return an error from the reconciler. That tradeoff didn't seem worth it. Since I'm not doing that anymore, I can drop the custom error type.